src/game.cpp

```diff
@@ -5954,16 +5954,23 @@ void game::monmove()
 
 void game::do_blast( const tripoint &p, const int power, const int radius, const bool fire )
 {
+    // TODO: Rewrite this to use some sort of a flood fill, to keep it from damaging stuff
+    // on the other side of a wall and more importantly the other side of floor/ceiling
     int dam;
-    for( auto &&t : m.points_in_radius( p, radius, 0 ) ) {
+    // Arbitrarily cut z-radius by a factor of 4
+    const int z_radius = m.has_zlevels() && debug_mode ? radius / 4: 0;
+    for( auto &&t : m.points_in_radius( p, radius, z_radius ) ) {
         if( t == p ) {
             dam = 3 * power;
         } else {
             dam = 3 * power / rl_dist( p, t );
         }
         m.smash_items(t, dam);
-        m.bash( t, dam );
-        m.bash( t, dam ); // Double up for tough doors, etc.
+        // Only smash down through floors if we have 1 "extra" z-level of radius
+        // But smash up through ceilings even if we don't have that extra z-level
+        const bool smash_floors = t.z > p.z - z_radius;
+        m.bash( t, dam, false, false, nullptr, smash_floors );
```

Comment:
Yeah, that's what I wanted to eventually put there.
The flood fill could also be more complex than just regular flood->bash->damage and allow some sort of "concentration" or shaped charge kind of behavior, to allow properly blasting through walls.

> This works roughly as expected if I just pass true to map::bash() instead of smash_floors.

I wanted the bottom-most level not to have `bash_floor` applied, because with `bash_floor` a level below the current bashed level can be cleared too. So if the `bash_floor` was always true, the explosion could damage walls but not creatures on the lowest level it would affect.